File: workspace_go/services/triggers/src/shared/configuration/application/configMap.go
package configMap

import (
	config "github.com/Mapex-Solutions/mapexGoKit/microservices/config"
)

// DefaultConfiguration defines all configuration keys for the Triggers service.
//
// This configuration follows the centralized pattern from MapexOS service,
// providing a single source of truth for all service configuration.
//
// Configuration sources (in priority order):
//  1. Environment variables
//  2. Default values defined here
var DefaultConfiguration = []config.ConfigDefinition{

	/** HTTP Server Configuration */
	{Key: "http_port", Env: "HTTP_PORT", Type: "int", Default: 5006},
	{Key: "http_address", Env: "HTTP_ADDRESS", Type: "string", Default: "0.0.0.0"},

	{Key: "service_name", Env: "SERVICE_NAME", Type: "string", Default: "triggers"},
	{Key: "service_version", Env: "SERVICE_VERSION", Type: "string", Default: "1.0.0"},

	/* MongoDB Configuration */
	{Key: "mongo_uri", Env: "MONGO_URI", Type: "string", Default: "mongodb://localhost:27017"},
	{Key: "mongo_database", Env: "MONGO_DATABASE", Type: "string", Default: "triggers"},
	{Key: "mongo_pool_size", Env: "MONGO_POOL_SIZE", Type: "int", Default: 10},
	{Key: "mongo_monitor_interval", Env: "MONGO_MONITOR_INTERVAL", Type: "int", Default: 10},

	/* Redis Configuration */
	{Key: "redis_host", Env: "REDIS_HOST", Type: "string", Default: "localhost"},
	{Key: "redis_port", Env: "REDIS_PORT", Type: "int", Default: 6379},
	{Key: "redis_username", Env: "REDIS_USERNAME", Type: "string", Default: ""},
	{Key: "redis_password", Env: "REDIS_PASSWORD", Type: "string", Default: ""},
	{Key: "redis_db", Env: "REDIS_DB", Type: "int", Default: 0},
	{Key: "redis_shared_db", Env: "REDIS_SHARED_DB", Type: "int", Default: 5},

	/** NATS Configuration */
	{Key: "nats_url", Env: "NATS_URL", Type: "string", Default: "nats://localhost:4222"},
	{Key: "nats_username", Env: "NATS_USERNAME", Type: "string", Default: "service"},
	{Key: "nats_password", Env: "NATS_PASSWORD", Type: "string", Default: "service_secret", Sensitive: true},
	{Key: "nats_client_name", Env: "NATS_CLIENT_NAME", Type: "string", Default: "triggers-service"},
	{Key: "nats_batch_size", Env: "NATS_BATCH_SIZE", Type: "int", Default: 500},
	{Key: "nats_fetch_timeout", Env: "NATS_FETCH_TIMEOUT", Type: "int", Default: 1},

	/** Executor Configuration */
	{Key: "trigger_executor_workers", Env: "TRIGGER_EXECUTOR_WORKERS", Type: "int", Default: 50},

	/** Metrics Configuration */
	{Key: "metrics_go_collector", Env: "METRICS_GO_COLLECTOR", Type: "bool", Default: true},
	{Key: "metrics_process_collector", Env: "METRICS_PROCESS_COLLECTOR", Type: "bool", Default: true},

	/**
	* Defines the authentication strategy used by the platform (jwt or oauth2) and
	* how user roles are retrieved (token, db, or api).
	* Supports static JWT secrets (HS256) and external identity providers using JWKS (RS256).
	* Allows dynamic role extraction from tokens, local databases, or external APIs.
	 */
	{Key: "auth_strategy", Env: "AUTH_STRATEGY", Type: "string", Default: "jwt"},
	{Key: "auth_secret", Env: "AUTH_SECRET", Type: "string", Default: "a-string-secret-at-least-256-bits-long", Sensitive: true},
	{Key: "auth_jwks_url", Env: "AUTH_JWKS_URL", Type: "string", Default: ""},
	{Key: "auth_algorithm", Env: "AUTH_ALGORITHM", Type: "string", Default: "HS256"},

	{Key: "auth_roles_source", Env: "AUTH_ROLES_SOURCE", Type: "string", Default: "token"}, // token | db | api
	{Key: "auth_roles_path", Env: "AUTH_ROLES_PATH", Type: "string", Default: "roles"},
	{Key: "auth_roles_api_url", Env: "AUTH_ROLES_API_URL", Type: "string", Default: ""},

	/**
	*	My apiKey for authentication
	* Please replace this with your own apiKey for authentication
	* This api Is used for internal communication between services
	 */
	{Key: "internal_api_key", Env: "INTERNAL_API_KEY", Type: "string", Default: "5230c2e2-e245-468d-89e8-94154cf520d0", Sensitive: true},

	/** Environment administration */
	{Key: "go_env", Env: "GO_ENV", Type: "string", Default: "dev"},

	// log_level overrides the default log level (debug, info, warn, error).
	// When empty, defaults to debug (dev) or info (production).
	{Key: "log_level", Env: "LOG_LEVEL", Type: "string", Default: ""},

	// ctx_timeout is used by the global middleware to establish a request timeout in seconds
	// value on the context, which is then passed down to all handlers.
	{Key: "ctx_timeout", Env: "CTX_TIMEOUT", Type: "int", Default: 4},

	/** Permission Middleware Configuration */
	{Key: "mapexos_url", Env: "MAPEXOS_URL", Type: "string", Default: "http://localhost:5000"},

	/** Assets Service Configuration for internal API calls */
	{Key: "assets_url", Env: "ASSETS_URL", Type: "string", Default: "http://localhost:5002"},
}

File: workspace_js/services/js-executor/src/shared/configuration/application/configMap.ts
import { ConfigDefinition } from '@mapexos/microservices';

// InitConfig initializes the configuration module with the provided definitions.
export const defaultConfiguration: ConfigDefinition[] = [

	/** HTTP Server Configuration */
	{ key: 'http_port', env: 'HTTP_PORT', type: 'int', default: 8000 },
	{ key: 'http_address', env: 'HTTP_ADDRESS', type: 'string', default: '0.0.0.0' },

	{ key: 'service_name', env: 'SERVICE_NAME', type: 'string', default: 'js-executor' },
	{ key: 'service_version', env: 'SERVICE_VERSION', type: 'string', default: '1.0.0' },

	/* Redis Configuration */
	{ key: 'redis_host', env: 'REDIS_HOST', type: 'string', default: 'localhost' },
	{ key: 'redis_port', env: 'REDIS_PORT', type: 'int', default: 6379 },
	{ key: 'redis_username', env: 'REDIS_USERNAME', type: 'string', default: '' },
	{ key: 'redis_password', env: 'REDIS_PASSWORD', type: 'string', default: '' },
	{ key: 'redis_db', env: 'REDIS_DB', type: 'int', default: 1 },

	/* Redis Lock Configuration */
	{ key: 'redis_lock_drift_factor', env: 'REDIS_LOCK_DRIFT_FACTOR', type: 'int', default: 0.01 },
	{ key: 'redis_lock_retry_count', env: 'REDIS_LOCK_RETRY_COUNT', type: 'int', default: 10 },
	{ key: 'redis_lock_retry_delay', env: 'REDIS_LOCK_RETRY_DELAY', type: 'int', default: 200 },
	{ key: 'redis_lock_retry_jitter', env: 'REDIS_LOCK_RETRY_JITTER', type: 'int', default: 200 },

	/** NATS Configuration */
	{ key: 'nats_url', env: 'NATS_URL', type: 'string', default: 'nats://localhost:4222' },
	{ key: 'nats_username', env: 'NATS_USERNAME', type: 'string', default: 'service' },
	{ key: 'nats_password', env: 'NATS_PASSWORD', type: 'string', default: 'service_secret', sensitive: true },
	{ key: 'nats_client_name', env: 'NATS_CLIENT_NAME', type: 'string', default: 'js-executor-service' },

	/**
	 * Defines the authentication strategy used by the platform (jwt or oauth2) and
	 * how user roles are retrieved (token, db, or api).
	 * Supports static JWT secrets (HS256) and external identity providers using JWKS (RS256).
	 * Allows dynamic role extraction from tokens, local databases, or external APIs.
	 */
	{ key: 'auth_strategy', env: 'AUTH_STRATEGY', type: 'string', default: 'jwt' },
	{ key: 'auth_secret', env: 'AUTH_SECRET', type: 'string', default: 'a-string-secret-at-least-256-bits-long', sensitive: true },
	{ key: 'auth_jwks_url', env: 'AUTH_JWKS_URL', type: 'string', default: '' },
	{ key: 'auth_algorithm', env: 'AUTH_ALGORITHM', type: 'string', default: 'HS256' },

	{ key: 'auth_roles_source', env: 'AUTH_ROLES_SOURCE', type: 'string', default: 'token' }, // token | db | api
	{ key: 'auth_roles_path', env: 'AUTH_ROLES_PATH', type: 'string', default: 'roles' },
	{ key: 'auth_roles_api_url', env: 'AUTH_ROLES_API_URL', type: 'string', default: '' },

	/**
	 *	My apiKey for authentication
	 * Please replace this with your own apiKey for authentication
	 * This api Is used for internal communication between services
	 */
	{ key: 'internal_api_key', env: 'INTERNAL_API_KEY', type: 'string', default: '5230c2e2-e245-468d-89e8-94154cf520d0', sensitive: true },

	/** Assets Service Configuration */
	{ key: 'assets_service_url', env: 'ASSETS_SERVICE_URL', type: 'string', default: 'http://localhost:5002' },

	/** Environment administration */
	{ key: 'node_env', env: 'NODE_ENV', type: 'string', default: 'dev' },

	/** Log level override (silent, fatal, error, warn, info, debug, trace).
	 *  When set, overrides the auto-derived level from NODE_ENV.
	 *  Use LOG_LEVEL=silent for performance testing. */
	{ key: 'log_level', env: 'LOG_LEVEL', type: 'string', default: '' },

	/** MinIO/S3 Configuration */
	{ key: 'minio_endpoint', env: 'MINIO_ENDPOINT', type: 'string', default: 'localhost:9000' },
	{ key: 'minio_access_key', env: 'MINIO_ACCESS_KEY', type: 'string', default: 'mapexos_admin', sensitive: true },
	{ key: 'minio_secret_key', env: 'MINIO_SECRET_KEY', type: 'string', default: 'mapexos_admin_secret_change_me', sensitive: true },
	{ key: 'minio_use_ssl', env: 'MINIO_USE_SSL', type: 'bool', default: false },
	{ key: 'minio_region', env: 'MINIO_REGION', type: 'string', default: 'us-east-1' },
	{ key: 'minio_assets_bucket', env: 'MINIO_ASSETS_BUCKET', type: 'string', default: 'mapex-assets' },
	{ key: 'minio_templates_bucket', env: 'MINIO_TEMPLATES_BUCKET', type: 'string', default: 'mapex-templates' },
	{ key: 'minio_bytecode_bucket', env: 'MINIO_BYTECODE_BUCKET', type: 'string', default: 'mapex-bytecode' },

	/** TieredCache Configuration (L0=RAM, L1=Disk, L2=MinIO) */
	{ key: 'cache_l0_max_size', env: 'CACHE_L0_MAX_SIZE', type: 'int', default: 268435456 },  // 256MB
	{ key: 'cache_l0_max_items', env: 'CACHE_L0_MAX_ITEMS', type: 'int', default: 100000 },
	{ key: 'cache_l0_ttl_seconds', env: 'CACHE_L0_TTL_SECONDS', type: 'int', default: 300 },  // 5min
	{ key: 'cache_l1_enabled', env: 'CACHE_L1_ENABLED', type: 'bool', default: true },
	{ key: 'cache_l1_dir', env: 'CACHE_L1_DIR', type: 'string', default: '/tmp/mapexos/cache' },
	{ key: 'cache_l1_max_size', env: 'CACHE_L1_MAX_SIZE', type: 'int', default: 10737418240 }, // 10GB
	{ key: 'cache_l1_ttl_seconds', env: 'CACHE_L1_TTL_SECONDS', type: 'int', default: 3600 }, // 1h

	/** TieredCache Fallback Configuration (HTTP API when L2 misses) */
	{ key: 'cache_fallback_timeout', env: 'CACHE_FALLBACK_TIMEOUT', type: 'int', default: 5000 }, // ms

	/** Bytecode Cache - Skip L0 (RAM) since Script Registry already caches in RAM */
	{ key: 'cache_bytecode_skip_l0', env: 'CACHE_BYTECODE_SKIP_L0', type: 'bool', default: true },

	/** CPU Limit — the SINGLE knob for auto-tuning.
	 *  All worker/chunk/batch values are derived from this.
	 *  Set to match your pod/container CPU limit. */
	{ key: 'cpu_limit', env: 'CPU_LIMIT', type: 'int', default: 4 },

	/** Piscina Worker Threads (0 = auto: CPU_LIMIT - 1, minimum 1) */
	{ key: 'piscina_workers', env: 'PISCINA_WORKERS', type: 'int', default: 0 },

	/** V8 Isolate Configuration (per worker thread) */
	{ key: 'isolate_memory_limit_mb', env: 'ISOLATE_MEMORY_LIMIT_MB', type: 'int', default: 32 },

	/** Worker Script Execution Timeout in ms */
	{ key: 'worker_script_timeout_ms', env: 'WORKER_SCRIPT_TIMEOUT_MS', type: 'int', default: 10000 },

	/** Recycle V8 context every N events to prevent memory leaks */
	{ key: 'context_recycle_interval', env: 'CONTEXT_RECYCLE_INTERVAL', type: 'int', default: 10000 },

	/** NATS Consumer Tuning (0 = auto from CPU_LIMIT) */
	{ key: 'nats_consumer_batch_size', env: 'NATS_CONSUMER_BATCH_SIZE', type: 'int', default: 0 },          // 0 = auto (CPU_LIMIT × 500)
	{ key: 'nats_consumer_fetch_timeout', env: 'NATS_CONSUMER_FETCH_TIMEOUT', type: 'int', default: 1000 }, // 1000ms = 1s fetch timeout
	{ key: 'nats_consumer_max_ack_pending', env: 'NATS_CONSUMER_MAX_ACK_PENDING', type: 'int', default: 0 }, // 0 = auto (batchSize × 2)

	/** Concurrency chunk size (0 = auto from PISCINA_WORKERS × 8) */
	{ key: 'concurrency_chunk_size', env: 'CONCURRENCY_CHUNK_SIZE', type: 'int', default: 0 },

	/** Events per piscina.run() call for batch workers (0 = auto: 500) */
	{ key: 'events_per_worker', env: 'EVENTS_PER_WORKER', type: 'int', default: 0 },
];
